hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/utils/EventBuffers.java

```diff
@@ -0,0 +1,120 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.sink.utils;
+
+import org.apache.hudi.common.util.Option;
+import org.apache.hudi.common.util.collection.Pair;
+import org.apache.hudi.sink.event.WriteMetadataEvent;
+
+import java.io.Serializable;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Stream;
+
+/**
+ * Utilities for coordinator event buffer.
+  */
+public class EventBuffers implements Serializable {
+  private static final long serialVersionUID = 1L;
+
+  // {checkpointId -> (instant, events)}
+  private final Map<Long, Pair<String, WriteMetadataEvent[]>> eventBuffers;
+
+  private EventBuffers(Map<Long, Pair<String, WriteMetadataEvent[]>> eventBuffers) {
+    this.eventBuffers = eventBuffers;
+  }
+
+  public static EventBuffers getInstance() {
+    return new EventBuffers(new ConcurrentHashMap<>());
+  }
+
+  /**
+   * Checks the buffer is ready to commit.
+   */
+  public static boolean allEventsReceived(WriteMetadataEvent[] eventBuffer) {
+    return Arrays.stream(eventBuffer)
+        // we do not use event.isReady to check the instant
+        // because the write task may send an event eagerly for empty
+        // data set, the even may have a timestamp of last committed instant.
+        .allMatch(event -> event != null && event.isLastBatch());
+  }
+
+  public WriteMetadataEvent[] addEventToBuffer(WriteMetadataEvent event) {
+    WriteMetadataEvent[] eventBuffer = this.eventBuffers.get(event.getCheckpointId()).getRight();
+    if (eventBuffer[event.getTaskID()] != null
+        && eventBuffer[event.getTaskID()].getInstantTime().equals(event.getInstantTime())) {
+      eventBuffer[event.getTaskID()].mergeWith(event);
+    } else {
+      eventBuffer[event.getTaskID()] = event;
+    }
+    return eventBuffer;
+  }
+
+  /**
+   * Returns existing bootstrap buffer or creates a new one.
+   */
+  public WriteMetadataEvent[] getOrCreateBootstrapBuffer(WriteMetadataEvent event, int parallelism) {
+    return this.eventBuffers.computeIfAbsent(event.getCheckpointId(),
+        ckpId -> Pair.of(event.getInstantTime(), new WriteMetadataEvent[parallelism])).getRight();
+  }
+
+  /**
+   * Cleans the task events triggered after or on the give event checkpoint ID.
+   */
+  public void cleanLegacyEvents(WriteMetadataEvent event) {
+    this.eventBuffers.entrySet().stream()
+        .filter(entry -> entry.getKey().compareTo(event.getCheckpointId()) >= 0)
```

Comment:
The cleanLegacyEvents operation may encounter performance issues in high-parallelism scenarios, such as with 6000 parallel tasks, due to repeated traversals of eventBuffers to identify the largest CheckpointID. To optimize this, could replacing the current data structure with an ordered one like TreeSet (which maintains sorted elements) be beneficial?

Reply:
like 
```
public void cleanLegacyEvents(WriteMetadataEvent event) {
    // tailMap O(log n + k)
    eventBuffers.tailMap(event.getCheckpointId(), true) 
                .values()
                .stream()
                .map(Pair::getRight)
                .forEach(eventBuffer -> resetBufferAt(eventBuffer, event.getTaskID()));
}
```